### ISSUE TITLE
Add support for attached properties 

### ISSUE THREAD

**Issue by stephenmdangelo** [qmlweb/qmlweb#297]
We kind of have support for attached properties for the "Component.onCompleted" signal, but it's handled as a special case. I'm currently working on adding the ability for registered constructors to have a "qmlAttachedProperties" method. I'm relying on the changes in #280 for allowing sets of attached properties on qualified types, but will make whatever changes are required of that gets changed based on whatever is accepted into qmlweb-parser. 

Adding support for referencing attached properties in JavaScript seems like it will be a little more work, as the attached property object isn't supposed to be created until it is first referenced. I'm going back o look into Proxy objects to dynamically create the object at access time. 

I'll push what I have in a PR likely on Monday, this is mostly a heads up that I'm working on this.

**Comment by akreuzkamp** [comment]
We do have support for attached properties, without any special handling. It's just broken ;)
I stumbled over it just today as well :)

Concretely, if you add a static method `getAttachedObject(obj)` to a class (e.g. `QMLComponent.getAttachedObject()`), it will be called as soon as you access an attached property. But the initialization still happens in `QMLEngine` and not on module initialization. It was forgotten to move it while merging. Have a look at QMLEngine.js:497, that's the code that needs to be moved. :)

**Comment by stephenmdangelo** [comment]
Are you sure it's 497? https://github.com/qmlweb/qmlweb/blob/master/src/engine/QMLEngine.js#L497

**Comment by stephenmdangelo** [comment]
The issue with how it is currently implemented is that the property name for the attaching type is added globally, which doesn't take into account a qualified module import (e.g. import QtQuick 2.0 as QQ; QQ.Component.onCompleted: ...).

I also don't like that every attaching type has to find a unique variable name on the item to which it is attaching to avoid conflicts (e.g. $Component). I'd like qmlweb to take care of that for developers. 

To avoid using a Proxy object, we could call "setupGetter" at item creation time with the appropriate property name for that context. This seems like it may negatively impact performance though, and seems like overkill for something like the "ListView" attaching type that isn't going to be used by 99% of the items created...

Also I'm mentioning "Proxy objects" like I know what I'm talking about. All I have is a vague idea that they may allow me to have a "get" function called when someone tries to access a missing property, I have to look into it further. I've currently just added one more case in "applyProperties" that will create the attached object when the property is set.

**Comment by akreuzkamp** [comment]
Sorry it's 582, I had the wrong branch checked out..

> The issue with how it is currently implemented is that the property name for the attaching type is added globally, which doesn't take into account a qualified module import (e.g. import QtQuick 2.0 as QQ; QQ.Component.onCompleted: ...).

That should be a matter of moving the initialization to the module init.

> I also don't like that every attaching type has to find a unique variable name on the item to which it is attaching to avoid conflicts (e.g. $Component). I'd like qmlweb to take care of that for developers.

This is something we indeed could and should change... When I implemented support for attached objects, I thought it was required for full compatibility, but it's not.

> To avoid using a Proxy object, we could call "setupGetter" at item creation time with the appropriate property name for that context. This seems like it may negatively impact performance though, and seems like overkill for something like the "ListView" attaching type that isn't going to be used by 99% of the items created...

This is - as you stated - overkill. But what we do, isn't so far from it. We add a getter for foo.AttachedType which will return the attached object. What we can do to improve this, is to add a .$attachedObjects object to every object, like we have for properties and save the attached objects there + doing the caching engine-internally.

Do you want to do that? Else I could implement it today, I need it as well and it should be done fast. As you prefer :)

**Comment by stephenmdangelo** [comment]
> Do you want to do that? Else I could implement it today, I need it as well and it should be done fast. As you prefer :)

I won't get to it until tomorrow, so by all means implement the solution you're thinking of. I can't quite picture how just moving the code that adds a getter to the QMLBaseObject prototype to module initialization will solve possible namespace clashes across files.

> What we can do to improve this, is to add a .$attachedObjects object to every object, like we have for properties and save the attached objects there + doing the caching engine-internally.

Yep, this is essentially the approach I took to cache the objects.

**Comment by akreuzkamp** [comment]
> I can't quite picture how just moving the code that adds a getter to the QMLBaseObject prototype to module initialization will solve possible namespace clashes across files.

You're right, it can't... I don't see a simple way to solve it. I consider adding the attached objects in the constructor a bad idea, as it would pollute the objects too much.. I'd accept the suboptimal situation for now, until we can rewrite code.

btw. I won't find the time to implement the fixes in the next 5 weeks, sorry (exams + vacation), feel free to get your hands dirty :)

**Comment by stephenmdangelo** [comment]
I'm looking into some sort of simple "transpiling" to support getting the attached objects.  I've added a pretty simple change to `qmlweb-parser` that wraps any dot-notation JavaScript code that references a `name` that matches `qml_is_element` with a call to `$QmlWebGetAttachedObject("ElementName")`. My current issue with this is how to re-generate the modified AST for the `src` that comes out of the parser.  I assume UglifyJS has a function I can call to do that, but we're only including the `parse-js` stuff at the moment? @ChALkeR , any advice on that would be appreciated.

We could then add a pretty simple `Object.prototype.$QmlWebGetAttachedObject` implementation that does a pass through for non-QObjects, and do whatever in-scope searching/creating of attached objects we need to for QObjects.  @akreuzkamp @ChALkeR Thoughts?
